spynnaker/pyNN/models/common/neuron_recorder.py

```diff
@@ -1,58 +1,520 @@
+                self._indexes[variable].sort()
+
+    def set_recording(self, variable, new_state, sampling_interval=None,
+                      indexes=None):
         if variable == "all":
-            for key in self._record.keys():
-                self._record[key] = new_state
-        elif variable in self._record:
-            self._record[variable] = new_state
+            for key in self._sampling_rates.keys():
+                self.set_recording(key, new_state, sampling_interval, indexes)
+        elif variable in self._sampling_rates:
+            if new_state:
+                self._turn_on_recording(variable, sampling_interval, indexes)
+            else:
+                self._turn_off_recording(variable, sampling_interval, indexes)
         else:
             msg = "Variable {} is not supported ".format(variable)
             raise fec_excceptions.ConfigurationException(msg)
 
-    def get_sdram_usage_in_bytes(self, variable, n_neurons,
-                                 n_machine_time_steps):
-        if self.is_recording(variable):
-            return recording_utils.get_recording_region_size_in_bytes(
-                n_machine_time_steps,  self.N_BYTES_PER_NEURON * n_neurons)
+    def get_buffered_sdram_per_record(self, variable, vertex_slice):
+        """
+        Returns the sdram used per record
+
+        :param variable:
+        :param vertex_slice:
+        :return:
+        """
+        n_neurons = self._count_recording_per_slice(variable, vertex_slice)
+        if n_neurons == 0:
+            return 0
+        if variable == SPIKES:
+            if n_neurons < vertex_slice.n_atoms:
+                # Indexing is used rather than gating to dettermine recording
+                # Non recoding neurons write to an extra slot
+                n_neurons += 1
+            out_spike_words = int(math.ceil(n_neurons / 32.0))
+            out_spike_bytes = out_spike_words * self.N_BYTES_PER_WORD
+            return self.N_BYTES_FOR_TIMESTAMP + out_spike_bytes
         else:
+            return self.N_BYTES_FOR_TIMESTAMP + \
+                        n_neurons * self.N_BYTES_PER_VALUE
+
+    def get_buffered_sdram_per_timestep(self, variable, vertex_slice):
+        """
+        Returns the sdram used per timestep
+
+        In the case where sampling is used it returns the average
+        for recording and none recording based on the recording rate
+
+        :param variable:
+        :param vertex_slice:
+        :return:
+        """
+        rate = self._sampling_rates[variable]
+        if rate == 0:
             return 0
 
-    def get_dtcm_usage_in_bytes(self):
-        return self.N_BYTES_PER_NEURON * sum(self._record.values())
+        data_size = self.get_buffered_sdram_per_record(
+            variable,  vertex_slice)
+        if rate == 1:
+            return data_size
+        else:
+            return data_size / rate
+
+    def get_sampling_overflow_sdram(self, vertex_slice):
```

Comment:
Not needed if get_buffered_sdram_per_timestep is changed to get_n_timesteps_in_buffer_space.

Reply:
get_n_timesteps_in_buffer_space is an UGLY approach that needs stripping out.

That assumes all processors are given an equal share of the remaining SDRAM and then the auto pause resume puts a break when the first one runs out.

Long term a better solution is to get the total sdram per timestep for ALL processor on the chip. Compute the max run time and then allocate the SDram per processor. But that is work for another day,

Reply:
I am fine with leaving this for another day - it just seemed like you were going out of your way to do per-timestep when this code would find it easier to work out the number of timesteps in a given buffer space (the code would be much clearer in that case). 

I have started a ticket for some possible ideas here: SpiNNakerManchester/SpiNNFrontEndCommon/293